<a href="https://colab.research.google.com/github/sakuronohana/my_datascience/blob/master/udemy/mlaz/Part%203%20-%20Classification/Section%2014%20-%20Logistic%20Regression/Python/logreg_practical_case_study.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Praktisches Fallbeispiel - Logistic regression

Im nachfolgenden Tutorial werden wir ein ML-Modell mittels der logistischen Regression aufbauen. Das das Ziel unsereres ML-Modells wird es sein, auf Basis eines Brustkrebsdatensatzen der Universität Wisconsin herauszufinden ob die Wahrscheinlichkeit für Brustkrebs besteht oder nicht.

Hierfür verwenden wir einen Datensatz der auf dem bekannten [ML-Repository UCI](https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29) abgelegt ist. Der Datensatz kommt aus dem 1992 ist aber für unsere Zwecke dennoch gut geeignet. Die Beschreibung des Datensatzes ist [hier](https://raw.githubusercontent.com/sakuronohana/my_datascience/master/udemy/mlaz/Part%203%20-%20Classification/Section%2014%20-%20Logistic%20Regression/Python/breast-cancer-wisconsin.txt) zu finden. Nachfolgend die wichtigsten Informationen über dem Datensatz.

* Datensatzmerkmale:  Multivariate
* Anzahl der Instanzen: 699
* Attributmerkmale: Ganze Zahl
* Anzahl der Attribute: 10 (unabhängige Variablen)
* Fehlende Werte?: Ja (16 fehlende Werte)

Attribute Information:

1. Sample code number: id number
2. Clump Thickness: 1 - 10
3. Uniformity of Cell Size: 1 - 10
4. Uniformity of Cell Shape: 1 - 10
5. Marginal Adhesion: 1 - 10
6. Single Epithelial Cell Size: 1 - 10
7. Bare Nuclei: 1 - 10
8. Bland Chromatin: 1 - 10
9. Normal Nucleoli: 1 - 10
10. Mitoses: 1 - 10
11. Class: (2 for benign (dt. gutartig), 4 for malignant (dt. bösartig))

### Beobachtungen aus Datensatzbeschreibung
1. Das Feature Class ist die abhängige Variable uns somit unser Label.
2. Wir müssen herausfinden ob Brustkrebs besteht oder nicht und haben es somti mit Classification und somit mit Supvervised Learning zu tun.
3. Die Wertskala der einzelnen Features (Attribute) deutet darauf hin, dass hier bereits eine Skalierung durchgeführt wurde. Wir also kein Features Scaling machen müssen.
4. Der Datensatz enhält 16 fehlende Werte. Wir müssen uns überlegen, ob wir damit klar kommen oder die entsprechenden Records rauslöschen oder ergänzen sollen.

# Import Libraries

Nachfolgend benötigen wir ein paar wichtige Python Bibliotheken

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Import Dataset


In [2]:
datloc = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/udemy/mlaz/Part%203%20-%20Classification/Section%2014%20-%20Logistic%20Regression/Python/dataset-breast-cancer.csv'
dataset = pd.read_csv(datloc)


# Preprocessing Dataset

In diesem Datensatz fehlen uns 16 Werte. Die Ersteller des Breast-Cancer Datensatzes haben die fehlenden Wert mit '?' ersetzt. Auch wenn wir mit 16 fehlenden Werten leben könnten, müssen wir die Fragezeichen loswerden, da wir ansonsten beim Training des Datensatzes Probleme bekommen. 

Nachfolgend löschen wir deshalb zuerst die Zeilen mit '?' und teilen den Datensatz erst danach in die unabhängigen Variablen (X) und abhängige Variablen (y).

In [3]:
dataset.drop( dataset[ dataset['7'] == '?' ].index , inplace=True)

In [4]:
# Bei der Selektion der Featrues für die unabhängigen Variabelen (X) lassen wir die
# erste Spalte mit den IDs und letzte Spalte mit den Labels weg
X = dataset.iloc[:,1:-1].values 
y = dataset.iloc[:,-1].values

# Split in Train- & Testset

In [5]:
from sklearn.model_selection import train_test_split

X_train,X_test, y_train,y_test = train_test_split(X,y,test_size = 0.25,random_state=0)

In [6]:
X_train.shape

(512, 9)

# Train the LogReg Model

In [7]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

# Predict the Testset results

In [9]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1),y_test.reshape(len(y_test),1)),1))

[[2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 4]
 [4 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]

# Visualize Pred. Result in Confusion Matrix

In [24]:
from sklearn.metrics import confusion_matrix, accuracy_score

print (confusion_matrix(y_test, y_pred))
tn,fp,fn,tp = confusion_matrix(y_test,y_pred).ravel()
print('Zuerst mal die postiven (True) Resultate:')
print ('true neg =',tn)
print ('true pos =',tp)
print('\n')
print('Nun noch die negativen (False) Resultate:')
print('false pos =',fp, '(auch Fehler 1.Art genannt)')
print('false neg =',fn, '(auch Fehler 2.Art genannt)')
print('\n')
print('Zum Abschluss noch die Genauigkeit des Modells')
print(accuracy_score(y_test,y_pred))


[[103   4]
 [  5  59]]
Zuerst mal die postiven (True) Resultate:
true neg = 103
true pos = 59


Nun noch die negativen (False) Resultate:
false pos = 4 (auch Fehler 1.Art genannt)
false neg = 5 (auch Fehler 2.Art genannt)


Zum Abschluss noch die Genauigkeit des Modells
0.9473684210526315


# Computing the accuracy with k-Fold Cross Validation

In [25]:
from sklearn.model_selection import cross_val_score
accuracy = cross_val_score(estimator=classifier, X = X_train, y=y_train, cv=10)
print('Accuracy: {:.2f}%'.format(accuracy.mean()*100))
print('Standard Deviation: {:.2f}%'.format(accuracy.std()*100))

Accuracy: 96.87%
Standard Deviation: 1.57%


Das Resultat der CV sagt nun folgendes aus:
Die Genauigkeit der 10 durchgeführten Stichproben weicht maximal 1.57% (+/-) von der Gesamtgenauigkeit von 96.87% des Models ab.

Wir haben somit eine sehr gutes Resultat bekommen bzw. die Genauigkeit der Vorhersagefähigkeit des Modells ist sehr gut. 